src/measure_theory/borel_hierarchy.lean

```diff
@@ -0,0 +1,440 @@
+augmented with `∅` and `univ` (intended to be a countable basis of a topology) and
+Σ⁰₀ is the empty family.
+-/
+inductive sigma0_pi0_rec {α : Type u} (s : set (set α)) :
+  ordinal.{u} → bool → set α → Prop
+| basic (x ∈ s) : sigma0_pi0_rec 0 tt x
+| empty         : sigma0_pi0_rec 0 tt ∅
+| univ          : sigma0_pi0_rec 0 tt univ
+| compl {i x}   : sigma0_pi0_rec i ff x → sigma0_pi0_rec i tt xᶜ
+| union {i} (f : ℕ → set α) (g : ℕ → ordinal.{u}) :
+    (∀ n, g n < i) → (∀ n, sigma0_pi0_rec (g n) tt (f n)) → sigma0_pi0_rec i ff (⋃ n, f n)
+
+/--
+The family of (boldface) Σ⁰ᵢ pointsets, which are countable unions of Π⁰ⱼ sets
+(given by the function `pointclasses.pi0` below) of smaller index. The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+-/
+def sigma0 : set (set α) := sigma0_pi0_rec s i ff
+
+/--
+The family of (boldface) Π⁰ᵢ pointsets, which are the complements of Σ⁰ᵢ sets
+(given by the function `pointclasses.sigma0` above). The parameter
+`s : set (set α)` is the family of sets from which the generation begins.
+
+When the ordinal argument is `0`, `pi0` returns the generating family `s`.
+-/
+def pi0 : set (set α) := sigma0_pi0_rec s i tt
+
+lemma sigma0_pi0_rec_def' {b : bool} :
+  sigma0_pi0_rec s i b = if b then pi0 s i else sigma0 s i :=
+by { unfold pi0 sigma0, cases b; refl }
+
+/-- Base case for `sigma0`. -/
+@[simp] lemma sigma0_zero : sigma0 s 0 = ∅ :=
+begin
+  unfold sigma0,
+  ext x, rw [mem_empty_iff_false, iff_false],
+  intro hx,
+  induction' hx with _ _ _ _ _ _ _ f g glt hf IH,
+  exact ordinal.not_lt_zero (g 0) (glt 0)
+ end
+
+/-- `sigma0` comprises unions of previous `pi0`. -/
+lemma sigma0_eq_Union_pi0:
+  sigma0 s i = set.range (λ (f : ℕ → ⋃ j (hij : j < i), pi0 s j), ⋃ n, (f n).1) :=
+begin
+  rcases classical.em (i=0) with rfl | hi; unfold sigma0, rw sigma0_pi0_rec_def',
+  { apply eq.symm, simp [range_eq_empty, ordinal.not_lt_zero] },
+  { ext x, split; intro hx,
+    { induction' hx with _ _ _ _ _ _ i f g glt hf,
+      existsi λn, (⟨f n, _⟩ : ↥⋃ j < i, pi0 s j),
+      { simp only [eq_self_iff_true] },
+      { rw mem_Union,
+      use g n,
+      rw mem_Union,
+      exact ⟨glt n, hf n⟩ } },
+    { cases hx with f hf,
+      rw ← hf,
+      dsimp only at hf,
+      choose g hg using λn, (mem_Union.mp (f n).property),
+      simp only [mem_Union,exists_prop] at hg,
+      apply sigma0_pi0_rec.union _ g,
+      exact λ n, (hg n).1,
+      unfold pi0 at hg,
+      exact λ n, (hg n).2 } }
+end
+
+lemma pi0_subset_sigma0 (hik : i < k) :
+  pi0 s i ⊆ sigma0 s k :=
+begin
+  simp only [sigma0_eq_Union_pi0,hik],
+  intros x hx,
+  apply mem_range.mpr,
+  have hxU : x ∈ ⋃ j < k, pi0 s j,
+  { simp only [mem_Union,exists_prop],
+    use i,
+    exact ⟨hik,hx⟩ },
+  existsi (λn : ℕ, (⟨x,hxU⟩ : ⋃ (j < k), pi0 s j)),
+  exact Union_const x
+end
+
+/-- `pi0` comprises complements of `sigma0` at the same level. -/
+lemma pi0_eq_compl_sigma0 (hi : ¬i = 0):
```

Comment:
```suggestion
lemma pi0_eq_compl_sigma0 (hi : i ≠ 0) :
```

Reply:
Here I preferred to put the simp-normal form of `i ≠ 0` in the hypothesis. Is that overkill?

Reply:
They are defeq, so it's mainly a notational issue. The standard spelling in mathlib is the one I suggested.

Reply:
It makes sense for it to be a `simp` lemma, since otherwise lemmas about equalities (e.g. `X = Y ↔ Z`) wouldn't fire on inequalities!

Reply:
I leave this discussion to you and the other experts! But this brings again this question:

> Here I preferred to put the simp-normal form of `i ≠ 0` in the hypothesis. Is that overkill?

Is this needed por `@[simp]` lemmas then? Or their assumptions are simplified before being applied?